pilot/pkg/networking/apigen/apigen.go

```diff
@@ -0,0 +1,165 @@
+// A first attempt added a 'sync' record at the end. Based on feedback and common use, a
+// different approach can be used - for large responses, we can mark the last one as 'hasMore'
+// by adding a field to the envelope.
+
+// Generate implements the generate method for high level APIs, like Istio config types.
+// This provides similar functionality with MCP and :8080/debug/configz.
+//
+// Names are based on the current resource naming in istiod stores.
+func (g *APIGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
+	resp := []*golangany.Any{}
+
+	// Note: this is the style used by MCP and its config. Pilot is using 'Group/Version/Kind' as the
+	// key, which is similar.
+	//
+	// The actual type in the Any should be a real proto - which is based on the generated package name.
+	// For example: type is for Any is 'type.googlepis.com/istio.networking.v1alpha3.EnvoyFilter
+	// We use: networking.istio.io/v1alpha3/EnvoyFilter
+	gvk := strings.SplitN(w.TypeUrl, "/", 3)
+	if len(gvk) == 3 {
+		rgvk := resource.GroupVersionKind{
+			Group:   gvk[0],
+			Version: gvk[1],
+			Kind:    gvk[2],
+		}
+		if w.TypeUrl == collections.IstioMeshV1Alpha1MeshConfig.Resource().GroupVersionKind().String() {
+			meshAny, err := gogotypes.MarshalAny(push.Mesh)
+			if err == nil {
+				resp = append(resp, &golangany.Any{
+					TypeUrl: w.TypeUrl,
+					Value:   meshAny.Value,
+				})
+			}
+			return resp
+		}
+
+		// TODO: what is the proper way to handle errors ?
+		// Normally once istio is 'ready' List can't return errors on a valid config -
+		// even if k8s is disconnected, we still cache all previous results.
+		// This needs further consideration - I don't think XDS or MCP transports
+		// have a clear recommendation.
+		cfg, err := push.IstioConfigStore.List(rgvk, "")
+		if err != nil {
+			log.Warnf("ADS: Error reading resource %s %v", w.TypeUrl, err)
+			return resp
+		}
+		for _, c := range cfg {
+			// Right now model.Config is not a proto - until we change it, mcp.Resource.
+			// This also helps migrating MCP users.
+
+			b, err := configToResource(&c)
+			if err != nil {
+				log.Warna("Resource error ", err, " ", c.Namespace, "/", c.Name)
+				continue
+			}
+			bany, err := gogotypes.MarshalAny(b)
+			if err == nil {
+				resp = append(resp, &golangany.Any{
+					TypeUrl: bany.TypeUrl,
+					Value:   bany.Value,
+				})
+			} else {
+				log.Warna("Any ", err)
+			}
+		}
+
+		// TODO: MeshConfig, current dynamic ProxyConfig (for this proxy), Networks
+
+		if w.TypeUrl == collections.IstioNetworkingV1Alpha3Serviceentries.Resource().GroupVersionKind().String() {
+			// Include 'synthetic' SE - but without the endpoints. Used to generate CDS, LDS.
+			// EDS is pass-through.
+			svcs := push.Services(proxy)
+			for _, s := range svcs {
+				c := external.ServiceToServiceEntry(s)
```

Comment:
Can you elaborate the use case here why we need convert k8s service to SE?

Reply:
That's how MCP worked. The MCP registry - which allegedly has other vendor implementations that can be consumed by Istio - is using ServiceEntry to represent endpoints and services, mapping whatever they use internally ( this is for non-k8s based Istio ). 

Eventually we need to split the endpoints to EndpointSlice and/or WorkloadEntry - but first step
is to have feature parity with MCP so we can deprecate MCP transport and unify on XDSv3/UDPA. We can improve after.